File: frontend/static/js/analytics.js
/**
 * Graph Analytics functionality
 */
class GraphAnalytics {
    constructor(cy, apiService) {
        this.cy = cy;
        this.apiService = apiService;
        this.highlightedElements = [];
    }

    // Helper functions for loading indicator
    showLoading() {
        document.getElementById('loading-overlay').style.display = 'flex';
    }

    hideLoading() {
        document.getElementById('loading-overlay').style.display = 'none';
    }

    /**
     * Find and highlight the shortest path between two nodes
     * @param {string} sourceId - ID of the source node
     * @param {string} targetId - ID of the target node
     * @param {Array<string>} relationshipTypes - Optional array of relationship types to consider
     */
    async findShortestPath(sourceId, targetId, relationshipTypes = null) {
        try {
            // Clear any previous highlights
            this.clearHighlights();

            // Show loading indicator
            this.showLoading();

            // Call the API to find the shortest path
            const result = await this.apiService.findShortestPath(sourceId, targetId, relationshipTypes);

            // Hide loading indicator
            this.hideLoading();

            if (!result.found) {
                alert(`No path found between ${sourceId} and ${targetId}`);
                return;
            }

            // Highlight the path in the graph
            this.highlightPath(result);

            // Show path information
            this.showPathInfo(result);

            return result;
        } catch (error) {
            this.hideLoading();
            console.error('Error finding shortest path:', error);
            alert(`Error finding shortest path: ${error.message}`);
        }
    }

    /**
     * Highlight a path in the graph
     * @param {Object} pathData - Path data from the API
     */
    highlightPath(pathData) {
        // First, make all elements transparent
        this.cy.elements().addClass('faded');

        // Collect elements to highlight
        const nodesToHighlight = [];
        const edgesToHighlight = [];

        // Add nodes to highlight
        pathData.nodes.forEach(node => {
            const cyNode = this.cy.getElementById(node.id);
            if (cyNode.length > 0) {
                nodesToHighlight.push(cyNode);
            }
        });

        // Add edges to highlight
        pathData.relationships.forEach(rel => {
            const sourceNode = this.cy.getElementById(rel.source_id);
            const targetNode = this.cy.getElementById(rel.target_id);

            if (sourceNode.length > 0 && targetNode.length > 0) {
                const edges = sourceNode.edgesTo(targetNode);
                if (edges.length > 0) {
                    edgesToHighlight.push(edges);
                }
            }
        });

        // Highlight the elements
        const elementsToHighlight = this.cy.collection();
        nodesToHighlight.forEach(node => elementsToHighlight.merge(node));
        edgesToHighlight.forEach(edge => elementsToHighlight.merge(edge));

        elementsToHighlight.removeClass('faded').addClass('highlighted');

        // Store highlighted elements for later cleanup
        this.highlightedElements = elementsToHighlight;

        // Fit the view to the highlighted elements
        this.cy.fit(elementsToHighlight, 50);
    }

    /**
     * Show path information in a modal or panel
     * @param {Object} pathData - Path data from the API
     */
    showPathInfo(pathData) {
        // Create HTML content for the path information
        let content = `
            <div class="path-info">
                <h5>Shortest Path</h5>
                <p><strong>Path Length:</strong> ${pathData.path_length}</p>
                <p><strong>Nodes:</strong> ${pathData.nodes.length}</p>
                <p><strong>Path:</strong></p>
                <ol class="path-nodes">
        `;

        // Add nodes to the path
        pathData.nodes.forEach((node, index) => {
            const nodeType = node.labels ? node.labels[0] : 'Unknown';
            const nodeName = node.name || node.id;

            content += `<li>${nodeName} (${nodeType})</li>`;

            // Add relationship if not the last node
            if (index < pathData.relationships.length) {
                const rel = pathData.relationships[index];
                content += `<li class="path-relationship">${rel.type}</li>`;
            }
        });

        content += `
                </ol>
            </div>
        `;

        // Show the content in the node details panel
        document.getElementById('node-details-title').textContent = 'Shortest Path';
        document.getElementById('node-details-content').innerHTML = content;
        document.getElementById('node-details').classList.add('visible');
    }

    /**
     * Find and highlight transaction clusters
     * @param {number} minClusterSize - Minimum number of transactions in a cluster
     * @param {number} maxDistance - Maximum distance between transactions
     */
    async findTransactionClusters(minClusterSize = 2, maxDistance = 2) {
        try {
            // Clear any previous highlights
            this.clearHighlights();

            // Show loading indicator
            this.showLoading();

            // Call the API to find transaction clusters
            const clusters = await this.apiService.getTransactionClusters(minClusterSize, maxDistance);

            // Hide loading indicator
            this.hideLoading();

            if (clusters.length === 0) {
                alert('No transaction clusters found with the specified parameters.');
                return;
            }

            // Show clusters information
            this.showClustersInfo(clusters);

            return clusters;
        } catch (error) {
            this.hideLoading();
            console.error('Error finding transaction clusters:', error);
            alert(`Error finding transaction clusters: ${error.message}`);
        }
    }

    /**
     * Show clusters information in a modal or panel
     * @param {Array} clusters - Array of cluster data from the API
     */
    showClustersInfo(clusters) {
        // Create HTML content for the clusters information
        let content = `
            <div class="clusters-info">
                <h5>Transaction Clusters</h5>
                <p><strong>Number of Clusters:</strong> ${clusters.length}</p>
                <div class="clusters-list">
        `;

        // Add clusters to the list
        clusters.forEach((cluster, index) => {
            content += `
                <div class="cluster-item">
                    <h6>Cluster ${index + 1} (${cluster.size} transactions)</h6>
                    <p><strong>Center:</strong> ${cluster.center_transaction.id}</p>
                    <p><strong>Transactions:</strong></p>
                    <ul>
            `;

            cluster.transactions.forEach(tx => {
                const amount = tx.amount || 0;
                const currency = tx.currency || 'USD';
                content += `<li>${tx.id} (${amount} ${currency})</li>`;
            });

            content += `
                    </ul>
                    <button class="btn btn-sm btn-outline-primary highlight-cluster" data-cluster-index="${index}">
                        Highlight Cluster
                    </button>
                </div>
            `;
        });

        content += `
                </div>
            </div>
        `;

        // Show the content in the node details panel
        document.getElementById('node-details-title').textContent = 'Transaction Clusters';
        document.getElementById('node-details-content').innerHTML = content;
        document.getElementById('node-details').classList.add('visible');

        // Add event listeners to highlight cluster buttons
        document.querySelectorAll('.highlight-cluster').forEach(button => {
            button.addEventListener('click', (event) => {
                const clusterIndex = parseInt(event.target.getAttribute('data-cluster-index'));
                this.highlightCluster(clusters[clusterIndex]);
            });
        });
    }

    /**
     * Highlight a transaction cluster in the graph
     * @param {Object} cluster - Cluster data from the API
     */
    highlightCluster(cluster) {
        // First, make all elements transparent
        this.cy.elements().addClass('faded');

        // Collect elements to highlight
        const elementsToHighlight = this.cy.collection();

        // Add transactions to highlight
        cluster.transactions.forEach(tx => {
            const cyNode = this.cy.getElementById(tx.id);
            if (cyNode.length > 0) {
                elementsToHighlight.merge(cyNode);

                // Also highlight connected edges
                elementsToHighlight.merge(cyNode.connectedEdges());
            }
        });

        // Highlight the elements
        elementsToHighlight.removeClass('faded').addClass('highlighted');

        // Store highlighted elements for later cleanup
        this.highlightedElements = elementsToHighlight;

        // Fit the view to the highlighted elements
        this.cy.fit(elementsToHighlight, 50);
    }

    /**
     * Show graph metrics in a modal or panel
     */
    async showGraphMetrics() {
        try {
            // Show loading indicator
            this.showLoading();

            // Call the API to get graph metrics
            const metrics = await this.apiService.getGraphMetrics();

            // Hide loading indicator
            this.hideLoading();

            // Create HTML content for the metrics
            let content = `
                <div class="metrics-info">
                    <h5>Graph Metrics</h5>
                    <div class="metrics-summary">
                        <div class="metric-item">
                            <span class="metric-value">${metrics.total_nodes}</span>
                            <span class="metric-label">Total Nodes</span>
                        </div>
                        <div class="metric-item">
                            <span class="metric-value">${metrics.relationship_count}</span>
                            <span class="metric-label">Relationships</span>
                        </div>
                    </div>

                    <div class="metrics-detail">
                        <h6>Node Types</h6>
                        <ul>
                            <li><strong>Users:</strong> ${metrics.user_count}</li>
                            <li><strong>Companies:</strong> ${metrics.company_count}</li>
                            <li><strong>Transactions:</strong> ${metrics.transaction_count}</li>
                        </ul>

                        <h6>Relationship Types</h6>
                        <ul>
            `;

            // Add relationship type counts
            Object.entries(metrics.relationship_type_counts).forEach(([type, count]) => {
                content += `<li><strong>${type}:</strong> ${count}</li>`;
            });

            content += `
                        </ul>

                        <h6>Most Connected Nodes</h6>
                        <ul>
            `;

            // Add most connected nodes
            metrics.most_connected_nodes.forEach(node => {
                content += `<li><strong>${node.name || node.id}:</strong> ${node.connection_count} connections</li>`;
            });

            content += `
                        </ul>
                    </div>
                </div>
            `;

            // Show the content in the node details panel
            document.getElementById('node-details-title').textContent = 'Graph Metrics';
            document.getElementById('node-details-content').innerHTML = content;
            document.getElementById('node-details').classList.add('visible');
        } catch (error) {
            this.hideLoading();
            console.error('Error getting graph metrics:', error);
            alert(`Error getting graph metrics: ${error.message}`);
        }
    }

    /**
     * Clear all highlights from the graph
     */
    clearHighlights() {
        this.cy.elements().removeClass('faded').removeClass('highlighted');
        this.highlightedElements = [];
    }
}

File: app/services/analytics.py
from app.database.connection import db
from typing import List, Dict, Any, Optional
from app.utils.serializers import serialize_neo4j_object

class GraphAnalyticsService:
    """Service for performing graph analytics operations"""

    @staticmethod
    def find_shortest_path(source_id: str, target_id: str, relationship_types: Optional[List[str]] = None) -> Dict[str, Any]:
        """
        Find the shortest path between two nodes in the graph
        
        Args:
            source_id: ID of the source node
            target_id: ID of the target node
            relationship_types: Optional list of relationship types to consider
                                If None, all relationship types are considered
        
        Returns:
            Dictionary containing the path information
        """
        # Build the relationship type filter if provided
        rel_filter = ""
        if relationship_types:
            rel_types = "|".join([f"`{rel}`" for rel in relationship_types])
            rel_filter = f"[:{rel_types}]"
        
        # Query to find the shortest path
        query = f"""
        MATCH (source {{id: $source_id}}), (target {{id: $target_id}}),
              p = shortestPath((source)-[{rel_filter}*]-(target))
        RETURN p, length(p) as path_length
        """
        
        parameters = {
            "source_id": source_id,
            "target_id": target_id
        }
        
        result = db.execute_query(query, parameters)
        
        if not result:
            return {
                "found": False,
                "message": f"No path found between {source_id} and {target_id}"
            }
        
        # Extract path information
        path = result[0]["p"]
        path_length = result[0]["path_length"]
        
        # Format the path for the response
        nodes = []
        relationships = []
        
        # Extract nodes from the path
        for node in path.nodes:
            nodes.append(serialize_neo4j_object(node))
        
        # Extract relationships from the path
        for rel in path.relationships:
            relationships.append({
                "type": rel.type,
                "source_id": rel.start_node["id"],
                "target_id": rel.end_node["id"],
                "properties": serialize_neo4j_object(rel)
            })
        
        return {
            "found": True,
            "path_length": path_length,
            "nodes": nodes,
            "relationships": relationships
        }
    
    @staticmethod
    def cluster_transactions(min_cluster_size: int = 2, max_distance: int = 2) -> List[Dict[str, Any]]:
        """
        Cluster transactions based on their connections
        
        Args:
            min_cluster_size: Minimum number of transactions in a cluster
            max_distance: Maximum distance between transactions to be considered in the same cluster
        
        Returns:
            List of transaction clusters
        """
        # Query to find transaction clusters
        query = f"""
        MATCH (t1:Transaction)
        CALL {{
            WITH t1
            MATCH (t1)-[*1..{max_distance}]-(t2:Transaction)
            WHERE t1.id <> t2.id
            RETURN t2
        }}
        WITH t1, collect(DISTINCT t2) AS connected_transactions
        WHERE size(connected_transactions) >= {min_cluster_size - 1}
        RETURN t1, connected_transactions
        """
        
        result = db.execute_query(query)
        
        # Process results to form clusters
        clusters = []
        processed_transactions = set()
        
        for record in result:
            t1 = record["t1"]
            t1_id = t1["id"]
            
            # Skip if this transaction is already in a cluster
            if t1_id in processed_transactions:
                continue
            
            # Create a new cluster
            cluster = {
                "center_transaction": serialize_neo4j_object(t1),
                "transactions": [serialize_neo4j_object(t1)],
                "size": 1 + len(record["connected_transactions"])
            }
            
            # Add connected transactions to the cluster
            for t2 in record["connected_transactions"]:
                t2_id = t2["id"]
                cluster["transactions"].append(serialize_neo4j_object(t2))
                processed_transactions.add(t2_id)
            
            # Add the center transaction to processed set
            processed_transactions.add(t1_id)
            
            # Add cluster to results
            clusters.append(cluster)
        
        return clusters
    
    @staticmethod
    def get_graph_metrics() -> Dict[str, Any]:
        """
        Calculate various metrics for the graph
        
        Returns:
            Dictionary containing graph metrics
        """
        # Query to get basic graph metrics
        query = """
        MATCH (n)
        OPTIONAL MATCH (u:User)
        OPTIONAL MATCH (t:Transaction)
        OPTIONAL MATCH (c:User {entity_type: 'company'})
        OPTIONAL MATCH ()-[r]-()
        RETURN 
            count(DISTINCT n) AS total_nodes,
            count(DISTINCT u) AS user_count,
            count(DISTINCT t) AS transaction_count,
            count(DISTINCT c) AS company_count,
            count(DISTINCT r) AS relationship_count
        """
        
        basic_metrics = db.execute_query(query)[0]
        
        # Query to get relationship type counts
        rel_query = """
        MATCH ()-[r]-()
        RETURN type(r) AS relationship_type, count(r) AS count
        ORDER BY count DESC
        """
        
        rel_counts = db.execute_query(rel_query)
        relationship_counts = {record["relationship_type"]: record["count"] for record in rel_counts}
        
        # Query to find most connected nodes
        connected_query = """
        MATCH (n)
        OPTIONAL MATCH (n)-[r]-()
        WITH n, count(r) AS connection_count
        ORDER BY connection_count DESC
        LIMIT 5
        RETURN n.id AS node_id, n.name AS node_name, labels(n) AS node_type, connection_count
        """
        
        connected_nodes = db.execute_query(connected_query)
        most_connected = [
            {
                "id": record["node_id"],
                "name": record.get("node_name", record["node_id"]),
                "type": record["node_type"][0] if record["node_type"] else "Unknown",
                "connection_count": record["connection_count"]
            }
            for record in connected_nodes
        ]
        
        return {
            "total_nodes": basic_metrics["total_nodes"],
            "user_count": basic_metrics["user_count"],
            "transaction_count": basic_metrics["transaction_count"],
            "company_count": basic_metrics["company_count"],
            "relationship_count": basic_metrics["relationship_count"],
            "relationship_type_counts": relationship_counts,
            "most_connected_nodes": most_connected
        }
